### PR TITLE
PersonnelRole Enum

### PR DESCRIPTION
**Pull request by Windchild292**
This PR implements PersonnelRole as an Enum, a grouping of role-related bugfixes, and adds the basis of the Dependent Role. There will be more changes in future for this role, and there _may_ be bugs for it. However, I believe the bugs are minimal as part of the larger implementation and they will be fixed in the near future as I move to a full rollout of the dependent role alongside #1886.

To review:
0.47.X: Do not.
Post-0.48.0: The goal is for this to be merged in for 0.49.0.

State:

- [x] PersonnelRole Enum
- [x] Dependent Role Baseline
- [x] PersonnelRole Enum Migration
- [x] PersonnelRole Enum-based Array Migration
- [x] Fix Unit Testing

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/personnel/Person.java [hunk 1]
```diff
@@ -2536,16 +2270,15 @@ public int hashCode() {
     }
 
     public int getExperienceLevel(boolean secondary) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/personnel/Person.java:2272]: I'm not certain the logic for this method should stay here or be moved to the PersonnelRole Enum given how much it relies on the latter, but the heavy reliance on Personnel is why I've kept it here.

File: MekHQ/src/mekhq/campaign/personnel/Person.java [hunk 2]
```diff
@@ -1080,123 +894,70 @@ public String getPrimaryRoleDesc() {
...
-            case T_VEHICLE_CREW:
-            case T_NONE:
-            default:
-                return KeyEvent.VK_UNDEFINED;
-        }
+        return getSecondaryRole().getName(isClanner());
     }
 
-    public boolean canPerformRole(int role) {
+    public boolean canPerformRole(PersonnelRole role) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/personnel/Person.java:904]: I'm not certain the logic for this method should stay here or be moved to the PersonnelRole Enum given how much it relies on the latter, but the heavy reliance on Personnel is why I've kept it here.

File: MekHQ/src/mekhq/campaign/personnel/Person.java [hunk 3]
```diff
@@ -700,156 +663,74 @@ public String getPortraitFileName() {
...
-                || (role == T_VEHICLE_CREW);
-    }
-
-    /**
-     * @param role A value that can be used for a person's primary or secondary role.
-     * @return     Whether the role is considered a support role
-     */
-    public static boolean isSupportRole(int role) {
-        return (role >= T_MECH_TECH) && (role < T_LAM_PILOT);
+    public boolean hasSupportRole(boolean excludeUnmarketable) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/personnel/Person.java:732]: Key note for review: the boolean has been flipped, so anyone using the former value will be doing so incorrectly.

**Review by NickAragua — CHANGES_REQUESTED** [review]
Nothing major. No need to do anything with the ternary operator comments, they're only there to illustrate a point about relative readability of code to different people.

File: MekHQ/src/mekhq/campaign/CampaignSummary.java [hunk 4]
```diff
@@ -93,9 +93,9 @@ public void updateInformation() {
             if (p.getHits() > 0) {
                 totalInjuries++;
             }
-            if (p.hasPrimaryCombatRole()) {
+            if (p.getPrimaryRole().isCombat()) {
                 totalCombatPersonnel++;
-            } else if (p.hasPrimarySupportRole(false)) {
+            } else {
```
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/campaign/CampaignSummary.java:98]: In the vast majority of cases they are. I've not, as per the overall description, finalized the dependent role, but am planning it being largely considered as support personnel.

File: MekHQ/src/mekhq/campaign/personnel/Person.java [hunk 5]
```diff
@@ -3788,11 +3379,9 @@ public void setEngineer(boolean b) {
      */
     public Money getRansomValue() {
         // MechWarriors and aero pilots are worth more than the other types of scrubs
-        if ((primaryRole == T_MECHWARRIOR) || (primaryRole == T_AERO_PILOT)) {
-            return MECHWARRIOR_AERO_RANSOM_VALUES.get(getExperienceLevel(false));
-        } else {
-            return OTHER_RANSOM_VALUES.get(getExperienceLevel(false));
-        }
+        return (getPrimaryRole().isMechWarriorGrouping() || getPrimaryRole().isAerospacePilot()
```
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/campaign/personnel/Person.java:3382]: I do for now, but will look at this specific example in the near future as I want to make the ransom values customizable.

File: MekHQ/src/mekhq/gui/CampaignGUI.java [hunk 6]
```diff
@@ -1240,20 +1240,20 @@ public boolean nagOutstandingScenarios() {
...
         }
-        
+
         return false;
     }
-   
-    private void hirePerson(java.awt.event.ActionEvent evt) {
+
+    private void hirePerson(ActionEvent evt) {
         int type = Integer.parseInt(evt.getActionCommand());
```
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/gui/CampaignGUI.java:1254]: Missed this in removal.

File: MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java [hunk 7]
```diff
@@ -1311,65 +1309,54 @@ private void loadGMToolsForPerson(Person person) {
...
-                }
-                if ((person.getPrimaryRole() == Person.T_DOCTOR) && (i == Person.T_MEDIC)) {
+        JMenuHelpers.addMenuIfNonEmpty(popup, menu);
+
+        menu = new JMenu(resourceMap.getString("changeSecondaryRole.text"));
+        for (final PersonnelRole role : roles) {
+            if (!role.isDependent()
+                    && (role.isNone() || (person.canPerformRole(role) && (person.getPrimaryRole() != role)))) {
+                // You can't be an astech if you are a tech, or a medic if you are a doctor
+                if ((person.getPrimaryRole().isTech() && role.isAstech())
```
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java:1332]: It is probably a good idea, but I'll fully check tomorrow.

File: MekHQ/src/mekhq/campaign/CampaignSummary.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/campaign/CampaignSummary.java:98]: This is also consistent with their current implementation in MekHQ.

File: MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Windchild292) [reply MekHQ/src/mekhq/gui/adapter/PersonnelTableMouseAdapter.java]: Folded it in, and expanded on a bunch of cases that both this and setPrimaryRole missed, plus removed the now useless setSecondaryRole checks

**Comment by Windchild292** [comment]
Ready for re-review @NickAragua.

**Review by NickAragua — APPROVED** [review]

**Pull request merged**